api/api/middleware/strict_auth_middleware.py

```diff
@@ -0,0 +1,21 @@
+from django.http import JsonResponse
+from rest_framework.status import HTTP_401_UNAUTHORIZED
+
+def strict_auth_middleware(get_response):
+    # Inner function to process each request
+    def middleware(request):
+        # Extract the Authorization header from the request
+        auth_header = request.headers.get('Authorization', None)
+        
+        # If the Authorization header is present
+        if auth_header:
+            # Check if the user is anonymous or authentication failed
+            if request.user.is_anonymous or request.auth is None:
```

Comment:
Because of how the OAuth2 DRF authenticator works, we can replace all of these checks with a check against `oauth2_error` on the request object:

https://github.com/jazzband/django-oauth-toolkit/blob/master/oauth2_provider/contrib/rest_framework/authentication.py#L30

The request attribute comes from this method, where you can see various error messages are turned as part of it:

https://github.com/jazzband/django-oauth-toolkit/blob/master/oauth2_provider/oauth2_validators.py#L219-L249

We haven't defined set scopes anywhere, so we don't need to worry about `insufficient_scope`, and can just assume the reason is for an invalid token. It'd be nice to return the error description as well, so users understand _why_ it's invalid.

Something along these lines is where I would start:

```py
oauth2_error = getattr(request, "oauth2_error", None)

if oauth2_error is None:
    return get_response(request)

raise AuthenticationFailed(
    detail=oauth2_error.get("error_description", None),
    code=oauth2_error.get("error", None),
)
```